### PR TITLE
Stronger FHIRPath Support

### PR DESCRIPTION
**Pull request by ms-k1ngk0ng**
# External FHIR Path Support

## Overview
Testing in FHIR often involves inspection of specific details of returned FHIR resources. FHIR Path is the format-agnostic language for pulling specific values out of a FHIR resource and several elements within TestScript support the provision of a FHIR Path expression, including in assertions and variable definitions. The Ruby-based FHIR Crucible implementation of FHIR path covers the basics, but does not support FHIR Path in full.  Several external tools do include more robust implementations of FHIR Path and other users/contributors to the FHIR Crucible project have decided to use those external engines rather than investing further in the FHIR Crucible implementation. While the TestScript-engine may be run in contexts where external tools are not available and so need to use the FHIR Crucible implementation, implementing optional support for an external FHIR path engine will allow further TestScript experimentation and potentially help to guide targeted improvements to the FHIR Crucible FHIR Path implementation.

## Requirements

1. Support for FHIR Path evaluation using the FHIR Crucible implementation
2. Support for FHIR Path evaluation using an external FHIR Path engine
3. Command-line configuration options to be able to decide which to use

## Proposed Approach

We propose to implement support for a single external FHIR Path implementation to allow for this experimentation with TestScripts not yet supported by the FHIR Crucible FHIR Path implementation. Specifically, we will use the Inferno [fhir-validator-wrapper](https://github.com/inferno-framework/fhir-validator-wrapper) project which includes a [FHIR Path evaluation API](https://github.com/inferno-framework/fhir-validator-wrapper/blob/master/rest-api.md#fhirpath-routes).

- Default = use internal FHIR Crucible FHIR Path implementation
- Options
  - useExternalFHIRPathEngine flag -> defaults to False. If true external FHIRPath evaluation will be used
  - ExternalFHIRPathEngineURL -> used only if useExternalFHIRPathEngine is True. defaults to the ONC Inferno public instance if not set and useExternalFHIRPathEngine is True. Used to direct to an alternate location, such as a local copy. Requests sent to [ExternalFHIRPathEngineURL]/[evaluate](https://github.com/inferno-framework/fhir-validator-wrapper/blob/master/rest-api.md#evaluate-a-fhirpath-expression).

## Example TestScripts
- https://github.com/fhir-crucible/testscript-engine/blob/mCODE/TestScripts/TestScript-us-mcode-profile-search-comorbidities.json - setup.action[1].assert.expression
- https://github.com/fhir-crucible/testscript-engine/blob/mCODE/TestScripts/TestScript-us-mcode-identify-cancer-patients-by-group.json - variable[0].expression

# Archived original WIP PR
Work-in-prorgess (WIP) PR to add support for external FHIRPath service. However, the default points to inferno, which is throwing a weird '500 Internal Server Error'. So, even though this PR is straightforward, leaving it as a WIP until the validator issue gets resolved on the Inferno end. Until that is resolved, can't really build a standardized service since the response format may change depending on the endpoint.

### CODE REVIEW

**Comment by jhlee-mitre** [comment]
Closing this request and archived as we don't continue adding new features at this point.
Later, we will consider merging all the features dealing with external services under one utility module.

**Comment by jhlee-mitre** [comment]
When pulled the PR, there are three materials to test the new fhirpath features:
example_bundle.json
example_bundle_entry.json (extract only entry from example_bundle.json)
testscript_fhirpath_test.json

Test scenario 1 (internal fhirpath evaluator)

`bundle exec testscript_engine execute --testscript_name testscript_fhirpath_test --testscript_path /spec/examples/testscript`

Expected result
```
STARTING TO TEST
         (✓) Expression: As expected, Expression equals ["true"]. Found ["true"].
         (✓) Expression: As expected, Expression equals ["393827"]. Found ["393827"].
         (✗) Expression: Expected Expression equals ["[393827]"], but found ["393827"].
         (✓) CompareToSourceId: As expected, CompareToSourceId equals ["8"]. Found ["8"].
         (✓) CompareToSourceId: As expected, CompareToSourceId equals ["[\"7216907\", {\"versionId\"=>\"1\", \"lastUpdated\"=>\"2022-12-20T15:24:50.183+00:00\", \"source\"=>\"#0Iq8V2GCfWbECpcY\"}, {\"status\"=>\"generated\"}, true, [{\"use\"=>\"official\", \"family\"=>\"Chalmers\", \"given\"=>[\"Peter\", \"James\"]}, {\"use\"=>\"official\", \"family\"=>\"Lee\", \"given\"=>[\"Jaehoon\", \"Bob\"]}], \"male\", \"1974-12-25\", \"Patient\"]"]. Found ["[\"7216907\", {\"versionId\"=>\"1\", \"lastUpdated\"=>\"2022-12-20T15:24:50.183+00:00\", \"source\"=>\"#0Iq8V2GCfWbECpcY\"}, {\"status\"=>\"generated\"}, true, [{\"use\"=>\"official\", \"family\"=>\"Chalmers\", \"given\"=>[\"Peter\", \"James\"]}, {\"use\"=>\"official\", \"family\"=>\"Lee\", \"given\"=>[\"Jaehoon\", \"Bob\"]}], \"male\", \"1974-12-25\", \"Patient\"]"].
         (✓) CompareToSourceId: As expected, CompareToSourceId equals ["[{\"use\"=>\"official\", \"family\"=>\"Chalmers\", \"given\"=>[\"Peter\", \"James\"]}, {\"use\"=>\"official\", \"family\"=>\"Lee\", \"given\"=>[\"Jaehoon\", \"Bob\"]}]"]. Found ["[{\"use\"=>\"official\", \"family\"=>\"Chalmers\", \"given\"=>[\"Peter\", \"James\"]}, {\"use\"=>\"official\", \"family\"=>\"Lee\", \"given\"=>[\"Jaehoon\", \"Bob\"]}]"].
      FINISHED TEST.
```

Test scenario 2 (external fhirpath evaluator)

`bundle exec testscript_engine execute --testscript_name testscript_fhirpath_test --ext_fhirpath "https://inferno.healthit.gov/validatorapi" --testscript_path /spec/examples/testscript`

Expected result

```
      STARTING TO TEST
            Initializing client with https://inferno.healthit.gov/validatorapi
            Configuring the client to use no authentication.
            POSTING: https://inferno.healthit.gov/validatorapi/evaluate?path=Bundle.total.toInteger()%20%3E=%20entry.count()
         (✗) Expression: Expected Expression equals ["true"], but found ["[true]"].
            Initializing client with https://inferno.healthit.gov/validatorapi
            Configuring the client to use no authentication.
            POSTING: https://inferno.healthit.gov/validatorapi/evaluate?path=Bundle.total
         (✗) Expression: Expected Expression equals ["393827"], but found ["[393827]"].
            Initializing client with https://inferno.healthit.gov/validatorapi
            Configuring the client to use no authentication.
            POSTING: https://inferno.healthit.gov/validatorapi/evaluate?path=Bundle.total
         (✓) Expression: As expected, Expression equals ["[393827]"]. Found ["[393827]"].
            Initializing client with https://inferno.healthit.gov/validatorapi
            Configuring the client to use no authentication.
            POSTING: https://inferno.healthit.gov/validatorapi/evaluate?path=Bundle.entry[0].resource.children().count()
            Initializing client with https://inferno.healthit.gov/validatorapi
            Configuring the client to use no authentication.
            POSTING: https://inferno.healthit.gov/validatorapi/evaluate?path=Patient.children().count()
         (✓) CompareToSourceId: As expected, CompareToSourceId equals ["[8]"]. Found ["[8]"].
            Initializing client with https://inferno.healthit.gov/validatorapi
            Configuring the client to use no authentication.
            POSTING: https://inferno.healthit.gov/validatorapi/evaluate?path=Bundle.entry[0].resource.children()
            Initializing client with https://inferno.healthit.gov/validatorapi
            Configuring the client to use no authentication.
            POSTING: https://inferno.healthit.gov/validatorapi/evaluate?path=Patient.children()
         (✓) CompareToSourceId: As expected, CompareToSourceId equals ["[\"7216907\", {\"versionId\"=>\"1\", \"lastUpdated\"=>\"2022-12-20T15:24:50.183+00:00\", \"source\"=>\"#0Iq8V2GCfWbECpcY\"}, {\"status\"=>\"generated\"}, true, {\"use\"=>\"official\", \"family\"=>\"Chalmers\", \"given\"=>[\"Peter\", \"James\"]}, {\"use\"=>\"official\", \"family\"=>\"Lee\", \"given\"=>[\"Jaehoon\", \"Bob\"]}, \"male\", \"1974-12-25\"]"]. Found ["[\"7216907\", {\"versionId\"=>\"1\", \"lastUpdated\"=>\"2022-12-20T15:24:50.183+00:00\", \"source\"=>\"#0Iq8V2GCfWbECpcY\"}, {\"status\"=>\"generated\"}, true, {\"use\"=>\"official\", \"family\"=>\"Chalmers\", \"given\"=>[\"Peter\", \"James\"]}, {\"use\"=>\"official\", \"family\"=>\"Lee\", \"given\"=>[\"Jaehoon\", \"Bob\"]}, \"male\", \"1974-12-25\"]"].
            Initializing client with https://inferno.healthit.gov/validatorapi
            Configuring the client to use no authentication.
            POSTING: https://inferno.healthit.gov/validatorapi/evaluate?path=Bundle.entry.resource.name
            Initializing client with https://inferno.healthit.gov/validatorapi
            Configuring the client to use no authentication.
            POSTING: https://inferno.healthit.gov/validatorapi/evaluate?path=Patient.name
         (✓) CompareToSourceId: As expected, CompareToSourceId equals ["[{\"use\"=>\"official\", \"family\"=>\"Chalmers\", \"given\"=>[\"Peter\", \"James\"]}, {\"use\"=>\"official\", \"family\"=>\"Lee\", \"given\"=>[\"Jaehoon\", \"Bob\"]}]"]. Found ["[{\"use\"=>\"official\", \"family\"=>\"Chalmers\", \"given\"=>[\"Peter\", \"James\"]}, {\"use\"=>\"official\", \"family\"=>\"Lee\", \"given\"=>[\"Jaehoon\", \"Bob\"]}]"].
      FINISHED TEST.
```

There are difference between internal and external evaluator to handle result. 
1. For internal evaluator, use the result as it is.
2. For external evaluator, peel off only "element" and wrap in Array. For example,

Returned object

```
[
    {
        "type": "integer",
        "element": 7
    }
]
```

Converted to:

```
[
    7
]
```

Returned object:

```
[
    {
        "type": "id",
        "element": "7216907"
    },
    {
        "type": "Meta",
        "element": {
            "versionId": "1",
            "lastUpdated": "2022-12-20T15:24:50.183+00:00",
            "source": "#0Iq8V2GCfWbECpcY"
        }
    }
]
```

Converted to:


```
[
    "7216907",
    {
            "versionId": "1",
            "lastUpdated": "2022-12-20T15:24:50.183+00:00",
            "source": "#0Iq8V2GCfWbECpcY"
        }
]
```

This type of conversion works to some extend, but at this point there is no way to completely remove the gap.

**Comment by jhlee-mitre** [comment]
Test scenario 1 (internal fhirpath evaluator)

`bundle exec testscript_engine execute --testscript_name testscript_fhirpath_test --testscript_path /spec/examples/testscript`

Expected result
```
      STARTING TO TEST
         (✗) Expression: Operator 'greaterThan' not supported in expression. Support only 'equals'
         (✓) Expression: As expected, Expression equals ["393827"]. Found ["393827"].
         (✓) CompareToSourceId: As expected, CompareToSourceId equals ["8"]. Found ["8"].
         (✓) CompareToSourceId: As expected, CompareToSourceId equals ["[\"7216907\", {\"versionId\"=>\"1\", \"lastUpdated\"=>\"2022-12-20T15:24:50.183+00:00\", \"source\"=>\"#0Iq8V2GCfWbECpcY\"}, {\"status\"=>\"generated\"}, true, [{\"use\"=>\"official\", \"family\"=>\"Chalmers\", \"given\"=>[\"Peter\", \"James\"]}, {\"use\"=>\"official\", \"family\"=>\"Lee\", \"given\"=>[\"Jaehoon\", \"Bob\"]}], \"male\", \"1974-12-25\", \"Patient\"]"]. Found ["[\"7216907\", {\"versionId\"=>\"1\", \"lastUpdated\"=>\"2022-12-20T15:24:50.183+00:00\", \"source\"=>\"#0Iq8V2GCfWbECpcY\"}, {\"status\"=>\"generated\"}, true, [{\"use\"=>\"official\", \"family\"=>\"Chalmers\", \"given\"=>[\"Peter\", \"James\"]}, {\"use\"=>\"official\", \"family\"=>\"Lee\", \"given\"=>[\"Jaehoon\", \"Bob\"]}], \"male\", \"1974-12-25\", \"Patient\"]"].
         (✓) CompareToSourceId: As expected, CompareToSourceId equals ["[{\"use\"=>\"official\", \"family\"=>\"Chalmers\", \"given\"=>[\"Peter\", \"James\"]}, {\"use\"=>\"official\", \"family\"=>\"Lee\", \"given\"=>[\"Jaehoon\", \"Bob\"]}]"]. Found ["[{\"use\"=>\"official\", \"family\"=>\"Chalmers\", \"given\"=>[\"Peter\", \"James\"]}, {\"use\"=>\"official\", \"family\"=>\"Lee\", \"given\"=>[\"Jaehoon\", \"Bob\"]}]"].
      FINISHED TEST.
```

Test scenario 2 (external fhirpath evaluator)

`bundle exec testscript_engine execute --testscript_name testscript_fhirpath_test --ext_fhirpath "https://inferno.healthit.gov/validatorapi" --testscript_path /spec/examples/testscript`

Expected result

```
 STARTING TO TEST
         (✗) Expression: Operator 'greaterThan' not supported in expression. Support only 'equals'
            POSTING: https://inferno.healthit.gov/validatorapi/evaluate?path=Bundle.total
         (✓) Expression: As expected, Expression equals ["393827"]. Found ["393827"].
            POSTING: https://inferno.healthit.gov/validatorapi/evaluate?path=Bundle.entry[0].resource.children().count()
            POSTING: https://inferno.healthit.gov/validatorapi/evaluate?path=Patient.children().count()
         (✓) CompareToSourceId: As expected, CompareToSourceId equals ["8"]. Found ["8"].
            POSTING: https://inferno.healthit.gov/validatorapi/evaluate?path=Bundle.entry[0].resource.children()
         (✗) Failed: FHIR Path returned (Array, 8 children) and can't compare to value (String)
            POSTING: https://inferno.healthit.gov/validatorapi/evaluate?path=Bundle.entry.resource.name
         (✗) Failed: FHIR Path returned (Array, 2 children) and can't compare to value (String)
      FINISHED TEST.
```

Changes:

1. When use FHIR path, operator must be "equals" or none, since value is defined as a String.
2. With that, a return from FHIR path should be a single String value. If it is multiple values, assertion fails with message of not supported. In case of Inferno, the engine takes "element" as a value, as long as it's single.

File: spec/examples/testscript/testscript_fhirpath_test.json [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+      "id": "02-PatientSearchDynamic",
+      "name": "Patient Search Dynamic",
+      "description": "Search for Patient resources using the user defined dynamic variables ${PatientSearchFamilyName} and ${PatientSearchGivenName} and validate response.",
+      "action": [
+        {
+          "assert": {
+            "description": "Confirm Bundle.total.toInteger() >= entry.count()",
+            "expression": "Bundle.total.toInteger() >= entry.count()",
+            "value": "true",
+            "operator": "greaterThan",
```
Comment (karlnaden) [inline spec/examples/testscript/testscript_fhirpath_test.json:101]: The operator here isn't needed, or should be `equals`.  You have a greater than operator in the expression, but you're asking the TestScript engine to compare the boolean result of the expression to the specified value `true` using equals.  Fine to not support operators other than `equals`, but in this case the operator should be `equals`.

File: spec/examples/testscript/testscript_fhirpath_test.json [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+            "compareToSourceId": "fixture-patient-bundle-entry",
+            "sourceId": "fixture-patient-bundle",
+            "warningOnly": false
+          }
+        },
+        {
+          "assert": {
+            "description": "Confirm Bundle.entry[0].resource.children() = Patient.children()",
+            "compareToSourceExpression": "Patient.children()",
+            "expression": "Bundle.entry[0].resource.children()",
```
Comment (karlnaden) [inline spec/examples/testscript/testscript_fhirpath_test.json:130]: This is a reasonable test case, but the difference in operation between the internal and external FHIR path engines is not appropriate.  If the external engine is not allowed to return a list, then the internal engine should not be able to do so either. The two should only have a different result if what they actually return is different and if you run these expressions on the internal and external engines, both return the same lists, with the external engine adding the wrapping as we know.  Easiest way to move forward is to add the restriction to the internal engine.

File: lib/testscript_engine/assertion.rb [hunk 3]
```diff
@@ -137,11 +134,23 @@ def content_type(assert)
...
+  def compare_to_source_id(assert)
+    resource = get_resource(assert.sourceId)
+    compare_to_source = get_resource(assert.compareToSourceId)
     raise AssertionException.new('No resource given by sourceId.', :fail) unless resource
+    raise AssertionException.new('No resource given by compareToSourceId.', :fail) unless compare_to_source
 
     received = evaluate_expression(assert.expression, resource)
     expected = determine_expected_value(assert)
-    compare("Expression", received, assert.operator, expected)
+    compare("CompareToSourceId", received.to_s, assert.operator, expected.to_s)
```
Comment (karlnaden) [inline lib/testscript_engine/assertion.rb:153]: The use of to_s here feels like a short cut that is probably right for now, but isn't always going to give you the correct results. For example, say you have a FHIRPath expression that evaluates to the boolean value `true` and another than evaluates to the string `"true"`, this logic won't be able to tell the difference. That said, I think ok to use this short-cut for now, but let's add an issue indicating that it is something to be resolved in the future by users interested in doing so.

**Review by karlnaden — CHANGES_REQUESTED** [review]
See comments for requested adjustments:
- different behavior between internal and external engines on same output
- change to example testscript
- issue to flag need for improved comparison

File: spec/examples/testscript/testscript_fhirpath_test.json (hunk 1 above)
Comment (jhlee-mitre) [reply spec/examples/testscript/testscript_fhirpath_test.json:101]: That's right. I put "greaterThan" intentionally to test it fails. (as in expected result) Sorry about lack of explanation.

File: lib/testscript_engine/assertion.rb (hunk 3 above)
Comment (jhlee-mitre) [reply lib/testscript_engine/assertion.rb:153]: Good point. I think it's trade-off by converting everything into String, practically it support real testing cases, on the other hands it rather mess up testscript especially if testers not aware of the conversion enough. 

I'm okay with either keeping this conversion or remove to_s. Or put it as another configurability?

**Review by karlnaden — APPROVED** [review]
moved the check for the FHIRpath returning an array to the assertion logic so that
- handled uniformly between internal and external FHIR path engines
- other features can accept a list of values
Clarified the example test script.

Expected results for both internal and external FHIR path on commands `bundle exec testscript_engine execute --testscript_name testscript_fhirpath_test --ext_fhirpath http://inferno.healthit.gov/validatorapi --testscript_path /spec/examples/testscript`  and `bundle exec testscript_engine execute --testscript_name testscript_fhirpath_test --testscript_path /spec/examples/testscript` are now
```
STARTING TO TEST
         (✗) Expression: Operator 'greaterThan' not supported in expression. Support only 'equals'
         (✓) Expression: As expected, Expression equals ["393827"]. Found ["393827"].
         (✓) CompareToSourceId: As expected, CompareToSourceId equals ["8"]. Found ["8"].
         (✗) CompareToSourceId: multiple values returned.
         (✗) CompareToSourceId: multiple values returned.
      FINISHED TEST.
```

**Review by jhlee-mitre — APPROVED** [review]
Approved. I am okay to keep "to_s" in returned values from FHIRpath if you're okay for now. Will merge into the main.

**Pull request merged**